### PR TITLE
Add score() and boost() + minor stuff

### PR DESCRIPTION
**Pull request by atombender**
First stab at document scoring. This is made harder by the fact that `score()` evaluates expressions that merely _contribute_ to scoring. So `match` predicates return true/false, not a score, so we have a few options:

* Specify a whole "parallel" expression specification that says "if used in scoring mode", the algorithm is: ..."
* Re-specify all predicate operators/functions as returning a `PredicateResult { result, score }` or a tuple `(result, score)` or similar.
* Handwave it away and say that `score()` evaluates the predicates and assigns a score based on whether the , but not exactly how. This is what I've done so far.

### CODE REVIEW

File: spec/Section 3 -- Execution.md [hunk 1]
```diff
@@ -76,13 +76,27 @@ Validate(expr):
...
+
+```
+* | score(a > 1)
+```
+
+should assign a score of 2.0 to any document where `a > 1`, and a score of 1.0 to any non-matching document.
+
+The scoring function for `match` is left as an implementation detail and not covered by this specification. For example, an implementation may choose to use a TD/IDF scoring function that uses the text corpus and language configuration for the given field to compute a text score.
+
+A boosted predicate simply adds the boost value to the score. For example, `boost(a > 1, 10)` would return `11` for any expression matching `a > 1`.
```
Comment (judofyr) [inline spec/Section 3 -- Execution.md:98]: As written here this mean that `boost(a > 1, 10)` will return `10` when _not_ matching `a > 1`. I don't think this is intended?

File: spec/Section 3 -- Execution.md [hunk 2]
```diff
@@ -76,13 +76,27 @@ Validate(expr):
...
+An expression is evaluated in a scope. You must successfully validate an expression before you attempt to evaluate it. Every expression type has their own evaluator function in their respective section in this specification (e.g. the evaluator of {ParenthesisExpression} is {EvaluateParenthesis()}).
 
 Evaluate(expr, scope):
 
 * Let {evaluator} be the evaluator of {expr}.
 * Return the result of {evaluator(scope)}.
 
+## Score evaluation
+
+When evaluating {score}, a predicate returning `true` should evaluate to the score `1.0`, and a `false` predicate should evaluate to the score `0.0`. For example, the expression:
```
Comment (judofyr) [inline spec/Section 3 -- Execution.md:88]: All predicates can return `null` as well. This should probably give `0` score?

File: spec/Section 3 -- Execution.md [hunk 3]
```diff
@@ -76,13 +76,27 @@ Validate(expr):
...
+
+```
+* | score(a > 1)
+```
+
+should assign a score of 2.0 to any document where `a > 1`, and a score of 1.0 to any non-matching document.
+
```
Comment (judofyr) [inline spec/Section 3 -- Execution.md:95]: We're missing `&&` and `||` here. As specified now `score(a > 1 || b > 2)` should only return a score of 1 or 0.

File: spec/Section 11 -- Pipe functions.md [hunk 4]
```diff
@@ -47,3 +47,58 @@ orderValidate(args):
...
+*[_type == "listing"] | score(body match "jacuzzi")
+```
+
+In this query, anything where `body match "jacuzzi"` returns true will be scored higher than other results. Multiple expressions can be used:
+
+```groq
+*[_type == "listing"] | score(body match "jacuzzi", bedrooms > 2, available && !inContract)
+```
+
+When multiple expressions are provided, the scores are merged into a single score for each result.
```
Comment (judofyr) [inline spec/Section 11 -- Pipe functions.md:65]: Maybe clarify that the scores are _added_ together?

File: spec/Section 11 -- Pipe functions.md [hunk 5]
```diff
@@ -47,3 +47,58 @@ orderValidate(args):
...
+
+Only predicate expressions that evaluate to a single boolean value may be used, including `boost()`.
```
Comment (judofyr) [inline spec/Section 11 -- Pipe functions.md:67]: I think we should either say something like "An implementation can further constrain which expressions are valid as a score expression [for optimization purposes]", _or_ provide a list of expressions which are an implementation _must_ support. The list could go under the "Scoring" section.

File: spec/Section 10 -- Functions.md [hunk 6]
```diff
@@ -213,3 +213,34 @@ stringValidate(args):
...
+
+The expression must be a predicate expressions that evaluates to a single boolean value. Any other result value is ignored.
+
+The value argument must be a number >= 0.
+
+The return value is the same as the input predicate. Internally, the scoring execution model uses the provided boost value to increase the computed score if the predicate matches.
+
+boost(args, scope):
+
+* Let {numNode} be the first element of {args}.
```
Comment (judofyr) [inline spec/Section 10 -- Functions.md:233]:
```suggestion
* Let {numNode} be the second element of {args}.
```

File: spec/Section 3 -- Execution.md (hunk 1 above)
Comment (atombender) [reply spec/Section 3 -- Execution.md]: That was implied, but I've improved the wording.

File: spec/Section 3 -- Execution.md (hunk 2 above)
Comment (atombender) [reply spec/Section 3 -- Execution.md]: Updated section.

**Review by judofyr — APPROVED** [review]
LGTM!

File: spec/Section 3 -- Execution.md [hunk 7]
```diff
@@ -76,13 +76,34 @@ Validate(expr):
...
+For logical expressions, the score is the sum of the clauses of the expression evaluates to `true`, otherwise 0.0. In other words:
+
+* `true && false` receives the score 0.0.
+* `true && true` receives the score 2.0.
+* `true || true` receives the score 2.0.
+* `true || false` receives the score 1.0.
+
+The scoring function for `match` is left as an implementation detail and not covered by this specification. For example, an implementation may choose to use a TD/IDF or similar text scoring function that uses the text corpus and language configuration for the given field to compute a text score.
```
Comment (j33ty) [inline spec/Section 3 -- Execution.md:103]: I think you meant `TF/IDF`.

File: spec/Section 3 -- Execution.md [hunk 8]
```diff
@@ -76,13 +76,34 @@ Validate(expr):
...
 Evaluate(expr, scope):
 
 * Let {evaluator} be the evaluator of {expr}.
 * Return the result of {evaluator(scope)}.
 
+## Score evaluation
+
+When evaluating {score}, a predicate returning `true` should have its score computed as 1.0, and all other values should receive a score of 0.0. All results involved in scoring start with a score of 1.0. The scores are evaluated once per result, and then added together. For example:
+
+```
```
Comment (j33ty) [inline spec/Section 3 -- Execution.md:90]: Minor: Missing language identifier for the code block.

**Review by j33ty — APPROVED** [review]

**Pull request merged**